### PR TITLE
Preserve constant values across union operations

### LINKED ISSUE #13804: Preserve constant values in union operations
### Is your feature request related to a problem or challenge?

When performing union operations in DataFusion, constant values are not preserved across partitions even when they are identical on both sides. This limits optimization opportunities.

### Describe the solution you'd like

Add constant value tracking to `ConstExpr` and preserve constant values across union operations when both sides have matching constants. This will enable better query optimization.

### Describe alternatives you've considered

Continue with current behavior where constants are not preserved across unions, but this misses optimization opportunities.

### Additional context

This enhancement will help with query optimization by maintaining constant value information across union operations when the values match on both sides.

### PR DESCRIPTION
**Pull request by gokselk**
## Which issue does this PR close?

Closes #13804.

## Rationale for this change

Currently, DataFusion doesn't preserve constant values across union operations even when both sides have the same constant value. This change enables better optimization by tracking and preserving constant values when they match.

## What changes are included in this PR?

- Added `value: Option<ScalarValue>` field to `ConstExpr`
- Added methods to get/set constant values
- Modified union operation logic to preserve matching constant values
- Updated equality comparison for `ConstExpr`
- Added tests for constant value preservation in unions

## Are these changes tested?

Yes, added new test case `test_union_constant_value_preservation` that verifies constant value preservation across unions.

## Are there any user-facing changes?

No user-facing changes. This is an internal optimization improvement.

### CODE REVIEW

**Comment by gokselk** [comment]
cc: @berkaysynnada @ozankabak

File: datafusion/physical-expr/src/equivalence/properties.rs
Comment (berkaysynnada) [inline datafusion/physical-expr/src/equivalence/properties.rs:1864]: When I remove this for loop, the tests don't fail. Can you check if they are really needed? If yes, can we write a test for that scenario also in this PR?

**Review by berkaysynnada — APPROVED** [review]
I have just one suggestion, otherwise LGTM

File: datafusion/physical-expr/src/equivalence/properties.rs
Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs]: I've tested it, and the constant removal loop appears to be redundant. I removed it in commit 291257f.

Comment (berkaysynnada) [reply datafusion/physical-expr/src/equivalence/properties.rs]: Can you understand why they were exist, and did they become redundant with this PR? They could do some work which does not appear at the tests. Maybe you can put some debug_asserts() to ensure we are not double counting (what the comment says)

Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs]: The constant removal loop was unnecessary even in the original code. The function already prevents double-counting by:

1. First collecting only the constants that exist in both LHS and RHS into a filtered `constants` vector 
2. Using only this filtered `constants` vector to create the final result via `with_constants()`
3. While `add_satisfied_orderings()` uses the original constant sets from LHS and RHS, this is correct because it's only checking if orderings from one side are satisfied in the other side. Having extra constants in the original sides doesn't affect this check

So modifying `lhs` and `rhs` by removing constants has no effect on the final result, as these modified properties aren't used in any way that would cause double-counting. The comment about "avoiding double counting" was likely added as a defensive measure.

File: datafusion/physical-expr/src/equivalence/properties.rs
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:3698]: Is there a way to crate an end to end `.slt` test that shows this behavior?

For example, a `EXPLAIN PLAN` where a Sort is optimized away after the constant value is propagated through the union?

**Review by alamb — COMMENTED** [review]
Thanks @gokselk  and @berkaysynnada 

I suggest we try to write an end to end sqllogictest for this query too.

File: datafusion/physical-expr/src/equivalence/properties.rs
Comment (berkaysynnada) [reply datafusion/physical-expr/src/equivalence/properties.rs:3698]: Good idea! I have one in my mind. Let me add it

Comment (berkaysynnada) [reply datafusion/physical-expr/src/equivalence/properties.rs:3698]: Hey @alamb, I have tried it but after thinking more, we actually need one more step in planner to experience an end-to-end difference. Now he have the knowledge, but we are not using it. 2 possible optimizations are which comes to my mind now:
Let's assume we have:
```
# Constant value tracking across union
query TT
explain
SELECT * FROM(
(
    SELECT * FROM aggregate_test_100 WHERE c1='a'
)
UNION ALL
(
    SELECT * FROM aggregate_test_100 WHERE c1='a'
))
ORDER BY c1
----
+   physical_plan
+   01)SortPreservingMergeExec: [c1@0 ASC NULLS LAST]
+   02)--UnionExec
+   03)----CoalesceBatchesExec: target_batch_size=2
+   04)------FilterExec: c1@0 = a
+   05)--------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
+   06)----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13], has_header=true
+   07)----CoalesceBatchesExec: target_batch_size=2
+   08)------FilterExec: c1@0 = a
+   09)--------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
+   10)----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13], has_header=true
```
1) At the top of the plan, we see an SPM. However, it can have a CoalescePartitionsExec instead. That would improve the performance for sure.
2) For the same query without an order by but with another outer filter, we will see another filter. However, we can actually remove that. This is another optimization, but can be observed pretty rarely rather than 1st one.

2nd one could be not really realistic, but the first one could be implemented without much effort with a few changes in replace_with_order_preserving_variants scope.

Comment (berkaysynnada) [reply datafusion/physical-expr/src/equivalence/properties.rs:3698]: Can you take a look at the first check @gokselk? It should take a few line changes in plan_with_order_preserving_variants() function. It should first look the order requirements, and if they are matched, then it would try to convert CoalescePartitionExec to SortPreservingMergeExec. But before that conversion, you can check across_partitions flag of the input constants, and if it is true, you can left the CoalescePartitionsExec as is.

Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs:3698]: > Can you take a look at the first check @gokselk? It should take a few line changes in plan_with_order_preserving_variants() function. It should first look the order requirements, and if they are matched, then it would try to convert CoalescePartitionExec to SortPreservingMergeExec. But before that conversion, you can check across_partitions flag of the input constants, and if it is true, you can left the CoalescePartitionsExec as is.

I looked into this optimization but it's not currently possible with just a few lines in `plan_with_order_preserving_variants()`. The challenge is that `PhysicalSortExpr` doesn't track value information, which prevents us from safely using `CoalescePartitionsExec` instead of `SortPreservingMergeExec`.
We would need to add value tracking to `PhysicalSortExpr` to implement this optimization.

Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs:3698]: > Can you take a look at the first check @gokselk? It should take a few line changes in plan_with_order_preserving_variants() function. It should first look the order requirements, and if they are matched, then it would try to convert CoalescePartitionExec to SortPreservingMergeExec. But before that conversion, you can check across_partitions flag of the input constants, and if it is true, you can left the CoalescePartitionsExec as is.

The issue is deeper than just checking `across_partitions`. Looking at `calculate_union()`, we're currently discarding all equivalence properties at UNION, including constants. We should preserve constants that are common across all UNION inputs.

For example, in:
```sql
(SELECT * FROM t WHERE x='a') 
UNION ALL 
(SELECT * FROM t WHERE x='a')
```
'x' remains constant='a' across the entire UNION, but we're losing this information. We should modify `calculate_union()` to preserve such common constants.

Would you like me to work on implementing this first?

Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs:3703]: > Can you take a look at the first check @gokselk? It should take a few line changes in plan_with_order_preserving_variants() function. It should first look the order requirements, and if they are matched, then it would try to convert CoalescePartitionExec to SortPreservingMergeExec. But before that conversion, you can check across_partitions flag of the input constants, and if it is true, you can left the CoalescePartitionsExec as is.

I've made changes to FilterExec for value extraction and added an initial SLT file. The query now shows `CoalescePartitionExec` in the output, so I think your suggested changes to `plan_with_order_preserving_variants()` might not be needed anymore. However, I'd appreciate your review to confirm this.

**Comment by ozankabak** [comment]
I wonder if we should change `across_partitions` to an `enum`; i.e.
```rust
enum PartitionValues {
    Uniform(Option<ScalarValue>),
    Heterogenous(Option<Vec<ScalarValue>>)
}
```

with `Uniform` meaning that all partitions have the same value given in the payload (if known), and `Heterogenous` meaning partitions can have different constant values (each of which is given in the vector, if known).

File: datafusion/physical-expr/src/equivalence/properties.rs
Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs:3703]: It appears that I broke some ORDER BY queries in my recent commits. I will investigate this further.

Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/properties.rs:3703]: To add more context, some tests are failing non-deterministically, which is why I didn't notice it beforehand.

Comment (ozankabak) [reply datafusion/physical-expr/src/equivalence/properties.rs:3703]: I'm not sure what the actual situation is w.r.t. those tests, but I'd advise to take a look at whether they were underspecified in the first place (i.e. the query itself may not be specifying a concrete output ordering, which could make the test flaky).

Do failing queries have top level ORDER BY clauses? If so, it is probably a bug that was introduced. Otherwise, maybe they were flaky in the first place.

File: datafusion/sqllogictest/test_files/union.slt
Comment (berkaysynnada) [inline datafusion/sqllogictest/test_files/union.slt:825]: This appears as SortPreservingMergeExec before these changes

**Comment by berkaysynnada** [comment]
I'm planning to merge this PR once the CI is all green

**Pull request merged**
